game.c

```diff
@@ -10,89 +10,98 @@
     getmaxyx(stdscr,row,col);
     row -= 1;
 
-    int dx = col/2 - lvl->player->x;
-    int dy = row/2 - lvl->player->y;
-
-    for (int x = 0; x < col; x++) {
-        for (int y = 0; y < row; y++) {
-            int xx = x - dx;
-            int yy = y - dy;
-            char display = UNSEEN;
-            // Positions which are part of the map
-            if (xx >= 0 && xx < lvl->width && yy >= 0 && yy < lvl->height) {
-                if (can_see(lvl, lvl->player, xx, yy)) {
-                    // Burning supercedes everything!
-                    if (lvl->chemistry[xx][yy]->elements[fire] > 0) {
-                        display = BURNING;
+    // Offset to keep player in center
+    int x_offset = (col / 2) - lvl->player->x;
+    int y_offset = (row / 2) - lvl->player->y;
+
+    for (int xx = 0; xx < col; xx++) {
+        for (int yy = 0; yy < row; yy++) {
+            // (xx,yy) are screen coordinates
+            // (x ,y ) are map coordinates
+            int x = xx - x_offset;
+            int y = yy - y_offset;
+
+            char icon = UNSEEN;
+
+            // Only draw squares which comprise the  map
+            if ((0 <= x && x < lvl->width) && (0 <= y && y < lvl->height)) {
+                // Only draw coords which the player can see
+                if (can_see(lvl, lvl->player, x, y)) {
+                    // Draw square status effects
+                    if (lvl->chemistry[x][y]->elements[fire] > 0) {
+                        icon = BURNING;
                         attron(COLOR_PAIR(RED));
+                    // Draw items
+                    } else if (lvl->items[x][y] != NULL) {
+                        icon = lvl->items[x][y]->item->display;
+                    // Draw floor
                     } else {
-                        // Draw items
-                        if (lvl->items[xx][yy] != NULL) {
-                            display = lvl->items[xx][yy]->item->display;
-                        } else {
-                            // Draw the square
-                            display = lvl->tiles[xx][yy];
-                        }
+                        icon = lvl->tiles[x][y];
                     }
                 }
-                if (display == UNSEEN) {
-                    display = lvl->memory[xx][yy];
+                // Fog of war
+                if (icon == UNSEEN) {
+                    icon = lvl->memory[x][y];
                     attron(COLOR_PAIR(YELLOW));
                 } else {
-                    lvl->memory[xx][yy] = display;
+                    lvl->memory[x][y] = icon;
                 }
             }
-            mvprintw(y, x, "%c", display);
-            // this should be a general unsetter, not this
+            mvprintw(yy, xx, "%c", icon);
+            //TODO this should be a general unsetter, not this
             attroff(COLOR_PAIR(YELLOW));
             attroff(COLOR_PAIR(RED));
         }
     }
+    // Mobs don't exist in the map. Coords exist on the mobs.
```

Comment:
Yah.

I put this here because so far the flow was "loops over each square, check what's there, draw that". Then the loop is over and a different loops starts for mobs. Not knowing the underlying structures, I would expect the first loop to also cover the mobs.

I don't actually know why mobs aren't on the map but haven't read the code yet.

Reply:
Since you need to operate on every mob every turn and they are expected to be sparse over the grid it makes sense to store them separately. You could store them on the grid like the items are too, it's kind of a toss up.